src/Pos/Wallet/Web/Server/Methods.hs

```diff
@@ -626,69 +626,86 @@ sendMoney sendActions passphrase moneySource dstDistr title desc = do
-            let fullHistory = taHistory <> cachedTxs
-                lenHistory = length taHistory
-                cached = drop (lenHistory - taCachedNum) taHistory
-            unless (null cached) $
-                updateHistoryCache
-                    accId
-                    taLastCachedHash
-                    taCachedUtxo
-                    (cached <> cachedTxs)
-
-            forM fullHistory $ addHistoryTx accId mempty mempty
-    pure (paginate cHistory, fromIntegral $ length cHistory)
+getFullWalletHistory :: WalletWebMode m => CId Wal -> m ([CTx], Word)
+getFullWalletHistory cWalId = do
+    accIds <- getWalletAccountIds cWalId
+    accAddrs <- concatMapM (getAccountAddrsOrThrow Ever) accIds
+    addrs <- mapM (decodeCIdOrFail . cwamId) accAddrs
+    cHistory <- do
+        (mInit, cachedTxs) <- transCache <$> getHistoryCache cWalId
+
+        -- TODO: Fix type param! Global type param.
+        TxHistoryAnswer {..} <- untag @WalletSscType getTxHistory addrs mInit
+
+        -- Add allowed portion of result to cache
+        let fullHistory = taHistory <> cachedTxs
+            lenHistory = length taHistory
+            cached = drop (lenHistory - taCachedNum) taHistory
+        unless (null cached) $
+            updateHistoryCache
+                cWalId
+                taLastCachedHash
+                taCachedUtxo
+                (cached <> cachedTxs)
+
+        forM fullHistory $ addHistoryTx cWalId mempty mempty
+    pure (cHistory, fromIntegral $ length cHistory)
   where
-    paginate = take defaultLimit . drop defaultSkip
-    defaultLimit = (fromIntegral $ fromMaybe 100 limit)
-    defaultSkip = (fromIntegral $ fromMaybe 0 skip)
+    transCache
+        :: Maybe (HeaderHash, Utxo, [TxHistoryEntry])
+        -> (Maybe (HeaderHash, Utxo), [TxHistoryEntry])
     transCache Nothing                = (Nothing, [])
     transCache (Just (hh, utxo, txs)) = (Just (hh, utxo), txs)
 
 -- FIXME: is Word enough for length here?
 searchHistory
     :: WalletWebMode m
-    => AccountId
-    -> Text
+    => Maybe (CId Wal)
+    -> Maybe CAccountId
     -> Maybe (CId Addr)
+    -> Maybe Text
     -> Maybe Word
     -> Maybe Word
     -> m ([CTx], Word)
-searchHistory accId search mAddrId skip limit = do
-    first (filter fits) <$> getHistory accId skip limit
+searchHistory mCWalId mCAccountId mAddrId mSearch mSkip mLimit = do
+    -- FIXME: searching when only AddrId is provided is not supported yet.
+    mAccountId <- mapM decodeCAccountIdOrFail mCAccountId
+    cWalId <- case (mCWalId, mAccountId) of
+        (Nothing, Nothing)        -> throwM errorSpecifySomething
+        (Just _, Just _)          -> throwM errorDontSpecifyBoth
+        (Just cWalId', Nothing)   -> pure $ cWalId'
+        (Nothing, Just accountId) -> pure $ aiWSId accountId
+    first (applySkipLimit . filter fits) <$> getFullWalletHistory cWalId
   where
-    fits ctx = txContainsTitle search ctx
+    fits ctx = maybe True (containsInTitle ctx) mSearch
```

Comment:
`maybe True ('txContainsTitle' ctx) mSearch`
it's not TH, it's just infix form of function )